### DISCUSSION TITLE
Can CompositeActivity be created using IActivityTypeProvider?

### DISCUSSION THREAD

**Discussion by tomy2105** [elsa-workflows/elsa-core#1290] [Q&A]
Related to #1122.
I tried creating activity derived from CompositeActivity using IActivityTypeProvider and it seems that my activity object is sometime created through callbacks provided by IActivityTypeProvider and sometimes directly.

The code that creates my activity object directly seems to come from WorkflowBlueprintMaterializer class:

            else if (typeof(CompositeActivity).IsAssignableFrom(activityType.Type))
            {
                var compositeActivity = (CompositeActivity) ActivatorUtilities.CreateInstance(_serviceProvider, activityType.Type);

So my question is, is it possible to use composite activity and create it using IActivityTypeProvider?

**Comment by sfmskywalker** [comment]
Good question. I have not tested this, so I don't know. I will try once I find some time.

**Comment by tomy2105** [comment]
It works :).
However its a bit puzzling as I said in initial post.

One instance of CompositeActivity derived class is created during service initialization (this one bypasses IActivityTypeProvider ), I think it is creating workflow blueprint at that time (not sure).
One more instance of CompositeActivity derived class seems to be created during workflow instance startup (this one bypasses IActivityTypeProvider ).
One more instance of CompositeActivity derived class is created during workflow instance execution, this one is done via IActivityTypeProvider.
There seems to be some more instances created unrelated to workflow execution, I think during bookmark indexing (those too bypasses IActivityTypeProvider)

So it works in general but it is a bit confusing and one must be aware that class needs to work both created directly and through IActivityTypeProvider)....

**In general it would be good to document when and for what are instances of activity class created and how they behave.**

**Comment by sfmskywalker** [comment]
Yeah the composite activity is a bit funny.

It acts as both a workflow builder as well as an activity that can execute, giving it two separate lifetimes, each of them living in a different dimension.

As you suggested in the other issue, we need to document this, including the various caveats to be aware of.

I intend to fix this good with Elsa 3, which will have a different API when it comes to building workflows and nested activities, where everything will feel more natural because the structure will actually be intrinsically hierarchical like Windows Workflow Foundation.

We will be able to do things like:

```csharp
public class MyWorkflow : Activity
{
   Start = new Sequence
   {
      new WriteLine("Hello!"),
      new MyComposite()
   }
}

public class MyComposite : Activity
{
   Start = new Sequence
   {
      new WriteLine("Hello from composite!"),
      new WriteLine("Later!"),
   }
}
```

In the above example, composites behave exactly the same as any other activity.
It also improves certain other things fundamentally for the engine.

Anyway, that is neither there nor here, so for the time being we will document the current behavior.
